src/open-hackathon-server/Kaiyuanshe.OpenHackathon.Server/Biz/TemplateRepoManagement.cs

```diff
@@ -0,0 +1,213 @@
+        Task DeleteTemplateRepoAsync([DisallowNull] string hackathonName, [DisallowNull] string templateRepoId, CancellationToken cancellationToken);
+    }
+
+    public class TemplateRepoManagement : ManagementClient<TemplateRepoManagement>, ITemplateRepoManagement
+    {
+        private static Regex regexGitHubRepoUrl = new Regex(@"(https:\/\/github\.com\/([A-Za-z0-9_\-]+)\/([A-Za-z0-9_\-]+))\/?");
+
+        #region Cache
+        private string CacheKeyByHackathon(string hackathonName)
+        {
+            return CacheKeys.GetCacheKey(CacheEntryType.TemplateRepo, hackathonName);
+        }
+
+        private void InvalidateCachedTemplateRepos(string hackathonName)
+        {
+            Cache.Remove(CacheKeyByHackathon(hackathonName));
+        }
+
+        private async Task<IEnumerable<TemplateRepoEntity>> GetCachedTemplateRepos(string hackathonName, CancellationToken cancellationToken)
+        {
+            string cacheKey = CacheKeyByHackathon(hackathonName);
+            return await Cache.GetOrAddAsync(cacheKey, TimeSpan.FromHours(6), (ct) =>
+            {
+                return StorageContext.TemplateRepoTable.ListByHackathonAsync(hackathonName, ct);
+            }, false, cancellationToken);
+        }
+        #endregion
+
+        private async Task<JObject> FetchGitHubRestApi(string url, CancellationToken cancellationToken = default)
+        {
+            using (var httpClient = HttpClientFactory.CreateClient())
+            using (var request = new HttpRequestMessage(HttpMethod.Get, url))
+            {
+                request.Headers.Add("Accept", "application/vnd.github+json");
+                // request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);
+
+                var response = await httpClient.SendAsync(request);
+                response.EnsureSuccessStatusCode();
+
+                using (var textReader = new StreamReader(await response.Content.ReadAsStreamAsync()))
+                using (var jsonReader = new JsonTextReader(textReader))
+                return (await JObject.LoadAsync(jsonReader, cancellationToken));
+            }
+        }
+
+        private async Task FetchGitHubInfo(TemplateRepo templateRepo, CancellationToken cancellationToken)
+        {
+            // Normalize the url.
+            var match = regexGitHubRepoUrl.Match(templateRepo.url);
+            if (!match.Success) {
+                return;
+            }
+            templateRepo.url = match.Groups[1].Value; // without trailing "/"
+            var owner = match.Groups[2].Value;
+            var repo = match.Groups[3].Value;
+
+            if (templateRepo.isFetched == true) {
+                return;
+            }
+
+            try {
+                // https://docs.github.com/en/rest/repos/repos#list-repository-languages
+                string url = string.Format("https://api.github.com/repos/{0}/{1}/languages", owner, repo);
```

Comment:
does this API allow anonymous call? If not, what's the required scope and how does OPH get it?

Reply:
Yes, I tested the API from local without cookie/token and it worked.

Reply:
I was wrong. GitHub API allows to access from browser, but it rejects when the HTTP request carrys minimal headers. We need a secret store first.